src/goto-programs/replace_function_bodies.cpp

```diff
@@ -190,26 +194,35 @@ class havoc_replace_function_bodiest : public replace_function_bodiest
     {
       if(
         parameter.type().id() == ID_pointer &&
-        !is_constant_or_has_constant_components(
-          parameter.type().subtype(), ns) &&
         std::regex_match(
           id2string(parameter.get_base_name()), parameters_to_havoc))
       {
-        // if (param != nullptr) { *param = nondet(); }
-        auto goto_instruction = add_instruction();
-        auto assign_instruction = add_instruction();
-        auto label_instruction = add_instruction();
-        goto_instruction->make_goto(
-          label_instruction,
-          equal_exprt(
-            symbol_exprt(parameter.get_identifier(), parameter.type()),
-            null_pointer_exprt(to_pointer_type(parameter.type()))));
-        assign_instruction->make_assignment();
-        assign_instruction->code = code_assignt(
-          dereference_exprt(
-            symbol_exprt(parameter.get_identifier(), parameter.type())),
-          side_effect_expr_nondett(parameter.type().subtype()));
-        label_instruction->make_skip();
+        if(
+          !is_constant_or_has_constant_components(
+            parameter.type().subtype(), ns))
+        {
+          // if (param != nullptr) { *param = nondet(); }
+          auto goto_instruction = add_instruction();
+          auto assign_instruction = add_instruction();
+          auto label_instruction = add_instruction();
+          goto_instruction->make_goto(
+            label_instruction,
+            equal_exprt(
+              symbol_exprt(parameter.get_identifier(), parameter.type()),
+              null_pointer_exprt(to_pointer_type(parameter.type()))));
+          assign_instruction->make_assignment();
+          assign_instruction->code = code_assignt(
+            dereference_exprt(
+              symbol_exprt(parameter.get_identifier(), parameter.type())),
+            side_effect_expr_nondett(parameter.type().subtype()));
+          label_instruction->make_skip();
+        }
+        else if(!parameter.type().subtype().get_bool(ID_C_constant))
+        {
+          warning() << "warning: not havoccing non-const pointer parameter '"
```

Comment:
Ah I see - this is why the tests are "wrong". I assume there is an issue? Probably worth ensuring this warning appears in the test and there is an appropriate comment on the issue to update the test.

Reply:
This is all intended behaviour for now, I had asked around what people thought the right behaviour should be and it was "just skip structs with const members for now". The other behaviour is also possible, would that be your preference then?

Reply:
On the face of it it seems more "correct" as I can easily write a method body for that method that violates that assert without wandering into undefined behaviour or even something particularly odd. That said, I'm not using this tool on a regular basis, it might be in the real world this is simply a more useful behaviour.